wa/instrumentation/fps.py

```diff
@@ -0,0 +1,165 @@
+                  which should not contribute to FPS calculation. 
+                  """),
+        Parameter('keep_raw', kind=bool, default=False,
+                  description="""
+                  If set to ``True``, this will keep the raw dumpsys output in
+                  the results directory (this is maily used for debugging)
+                  Note: frames.csv with collected frames data will always be
+                  generated regardless of this setting.
+                   """),
+        Parameter('crash_threshold', kind=float, default=0.7,
+                  description="""
+                  Specifies the threshold used to decided whether a
+                  measured/expected frames ration indicates a content crash.
+                  E.g. a value of ``0.75`` means the number of actual frames
+                  counted is a quarter lower than expected, it will treated as
+                  a content crash.
+
+                  If set to zero, no crash check will be performed.
+                  """),
+        Parameter('period', kind=float, default=2, constraint=lambda x: x > 0,
+                  description="""
+                  Specifies the time period between polling frame data in
+                  seconds when collecting frame data. Using a lower value
+                  improves the granularity of timings when recording actions
+                  that take a short time to complete.  Note, this will produce
+                  duplicate frame data in the raw dumpsys output, however, this
+                  is filtered out in frames.csv.  It may also affect the
+                  overall load on the system.
+
+                  The default value of 2 seconds corresponds with the
+                  NUM_FRAME_RECORDS in
+                  android/services/surfaceflinger/FrameTracker.h (as of the
+                  time of writing currently 128) and a frame rate of 60 fps
+                  that is applicable to most devices.
+                  """),
+        Parameter('force_surfaceflinger', kind=bool, default=False,
+                  description="""
+                  By default, the method to capture fps data is based on
+                  Android version.  If this is set to true, force the
+                  instrument to use the SurfaceFlinger method regardless of its
+                  Android version.
+                  """),
+    ]
+
+    def __init__(self, target, **kwargs):
+        super(FpsInstrument, self).__init__(target, **kwargs)
+        self.collector = None
+        self.processor = None
+        self._is_enabled = None
+
+    def setup(self, context):
+        use_gfxinfo = self.target.get_sdk_version() >= 23 and not self.force_surfaceflinger
+        if use_gfxinfo:
+            collector_target_attr = 'package'
+        else:
+            collector_target_attr = 'view'
+        collector_target = getattr(context.workload, collector_target_attr)
```

Comment:
Need to add `, None` here for the `if collector_target` to work (otherwise you get `AttributeError`)

Reply:
fixed.